providers/implementations/ciphers/cipher_chacha20_poly1305.c.in

```diff
@@ -103,181 +100,164 @@ static int chacha20_poly1305_get_params(OSSL_PARAM params[])
-{- produce_param_list('static', 'chacha20_poly1305_known_gettable_ctx_params',
-                      'static', 'chacha20_poly1305_get_ctx_params_find_pidx',
-                      (['CIPHER_PARAM_KEYLEN',                  'size_t'],
-                       ['CIPHER_PARAM_IVLEN',                   'size_t'],
-                       ['CIPHER_PARAM_AEAD_TAGLEN',             'size_t'],
-                       ['CIPHER_PARAM_AEAD_TAG',                'octet_string'],
-                       ['CIPHER_PARAM_AEAD_TLS1_AAD_PAD',       'size_t'],
-                      )); -}
-/* End of machine generated */
-
 static const OSSL_PARAM *chacha20_poly1305_gettable_ctx_params
     (ossl_unused void *cctx, ossl_unused void *provctx)
 {
-    return chacha20_poly1305_known_gettable_ctx_params;
+    return chacha20_poly1305_get_ctx_params_ettable;
 }
 
-/* Machine generated by util/perl/OpenSSL/paramnames.pm */
-{- produce_param_list('static', 'chacha20_poly1305_known_settable_ctx_params',
-                      'static', 'chacha20_poly1305_set_ctx_params_find_pidx',
-                      (['CIPHER_PARAM_KEYLEN',                  'size_t'],
-                       ['CIPHER_PARAM_IVLEN',                   'size_t'],
-                       ['CIPHER_PARAM_AEAD_TAG',                'octet_string'],
-                       ['CIPHER_PARAM_AEAD_TLS1_AAD',           'octet_string'],
-                       ['CIPHER_PARAM_AEAD_TLS1_IV_FIXED',      'octet_string'],
-                      )); -}
-/* End of machine generated */
+{- produce_param_decoder('chacha20_poly1305_set_ctx_params',
+                         (['CIPHER_PARAM_KEYLEN',             'keylen', 'size_t'],
+                          ['CIPHER_PARAM_IVLEN',              'ivlen',  'size_t'],
+                          ['CIPHER_PARAM_AEAD_TAG',           'tag',    'octet_string'],
+                          ['CIPHER_PARAM_AEAD_TLS1_AAD',      'aad',    'octet_string'],
+                          ['CIPHER_PARAM_AEAD_TLS1_IV_FIXED', 'fixed',  'octet_string'],
+                         )); -}
 
 static const OSSL_PARAM *chacha20_poly1305_settable_ctx_params(
         ossl_unused void *cctx, ossl_unused void *provctx
     )
 {
-    return chacha20_poly1305_known_settable_ctx_params;
+    return chacha20_poly1305_set_ctx_params_ettable;
 }
 
 static int chacha20_poly1305_set_ctx_params(void *vctx,
                                             const OSSL_PARAM params[])
 {
-    const OSSL_PARAM *p;
     size_t len;
     PROV_CHACHA20_POLY1305_CTX *ctx = (PROV_CHACHA20_POLY1305_CTX *)vctx;
     PROV_CIPHER_HW_CHACHA20_POLY1305 *hw =
         (PROV_CIPHER_HW_CHACHA20_POLY1305 *)ctx->base.hw;
+    struct chacha20_poly1305_set_ctx_params_st p;
 
     if (ossl_param_is_empty(params))
         return 1;
 
-    for (p = params; p->key != NULL; p++)
-        switch (chacha20_poly1305_set_ctx_params_find_pidx(p->key)) {
-        default:
-            break;
+    p = chacha20_poly1305_set_ctx_params_decoder(params);
```

Comment:
is_empty() check is kind of the same as p being all NULL I guess.. Adding counting for getters doesnt seem like it would be worth it..

Reply:
Only the supplied parameters are parsed and only once each, so no gain there.
Counting could be in addition to null checks but I doubt it would be a benefit.

Reply:
Note also that p cannot ever be NULL.  This function is returning a structure not a pointer.